File: lesson_5/permutations.py
# permutations

# ""
# ["abc","acb","bac","bca","cab","cba"]
# [""]
# "bc" - ["bc","cb"] - ["abc","acb"]
# "ac" - ["ac","ca"] - ["bac","bca"]
# "ab" - ["ab","ba"] - ["cab","cba"]

def permutations(a: str) -> [str]:
    if not a:
        return [""]
    else:
        answer = []
        for i in range(len(a)):
            reversal = permutations(a[:i] + a[i+1:])
            for j in range(len(reversal)):
                answer.append(a[i] + reversal[j])
        return answer

print(permutations("abc"))

File: recursion/permutations.go
package main

import "fmt"

func permutations(a string) []string {
	if len(a) == 0 {
		return []string{""}
	}
	var answer []string
	for i := 0; i < len(a); i++ {
		reversal := permutations(a[:i] + a[i+1:])
		for _, s := range reversal {
			answer = append(answer, string(a[i])+s)
		}
	}
	return answer
}

func main() {
	result := permutations("abc")
	fmt.Println(result)
}
